### DISCUSSION TITLE
Typed `ptr` property for arrays vs `[direct_array_access]`

### DISCUSSION THREAD

**Discussion by ntrel** [vlang/v#11818] [Ideas]
`[direct_array_access]` turns off bounds-checks for arrays for a whole function - this hides memory-unsafe behaviour. Instead we could make it easier to get an element pointer - what if we had a `ptr` array property for `[]T` that is `&T`,  unlike the `data voidptr` field. That avoids a pointer cast, and it's much cleaner for pointer indexing:

```v
arr[0] = 2 // looks like it should bounds check but might not
unsafe {(&int(arr.data))[0] = 2} // using .data
unsafe {arr.ptr[0] = 2} // proposed
```

`[direct_array_access]` is sometimes overkill, so making it easier to selectively avoid bounds checks inside a function is a win regardless.

Note also that `[direct_array_access]` can sometimes be avoided with for/in: https://github.com/vlang/v/pull/8857

**Comment by ntrel** [comment]
> #8857

Also #9059 removes **38** unnecessary uses of `[direct_array_access]` (due to literal indexing of fixed arrays, and also in some functions that only used array literals, no indexing). It also adds missing bounds checks for 6 uses of `[direct_array_access]` (in functions that weren't tagged `[unsafe]`).

**Comment by ntrel** [comment]
> It also adds missing bounds checks for 6 uses of [direct_array_access]

@spytheman I strongly think `[direct_array_access]` should require `unsafe {}` for each array indexing. Otherwise people will forget to check bounds as evidenced above. If `unsafe` was required, reviewers would be more likely to think 'why is it unsafe?' and probably realize that manual bounds checks were necessary for runtime indexing.

**Comment by dumblob** [comment]
Speaking of `[direct_array_access]` - would it make sense to issue a warning if in the given function no bounds check was needed to be escaped (e.g. because there is no array with runtime bounds checking or because there is no array whatsoever)?

**Comment by ntrel** [comment]
>  issue a warning if in the given function no bounds check was needed to be escaped

Yes, if we keep `[direct_array_access]` then it should check that it is used.

**Comment by penguindark** [comment]
Who use `[direct_array_access]` must know the risks, but if you need speed it is good. Totally safe vs speed is not acceptable in some applications. Like 3D programs.
I'm for mantain `[direct_array_access]`

**Comment by spytheman** [comment]
The goal of `[direct_array_access]` is that you can add it, or comment it very easily, without changing the code otherwise, so that an algorithm can be debugged as naturally as possible, then it can be added for speed, or commented again, when someone wants to debug it again.

Making an obtuse new syntax for getting a pointer instead will break that.

**Comment by spytheman** [comment]
> Yes, if we keep [direct_array_access] then it should check that it is used.

It would be an useful warning imho. @penguindark what do you think?

**Comment by spytheman** [comment]
The original code because of which `[direct_array_access]` was added is this one btw: https://github.com/vlang/v/blob/master/vlib/arrays/arrays.v#L68 . You can try rewriting that with the new proposed syntax, so we can compare the two versions for readability.

**Comment by penguindark** [comment]
I think that `[direct_array_access]` must remain with the double purpose of:
- not check bounds (its functional function)
- it is a clear warning of the intent of the user (its informative function)

Obviously we need to document well its behaviour

**Comment by spytheman** [comment]
> I think that `[direct_array_access]` must remain with the double purpose of:
> - not check bounds (its functional function)
> - it is a clear warning of the intent of the user (its informative function)
> 
> Obviously we need to document well its behaviour

Yes, but do you think there should be a warning/notice, if it is used for code that does not require it?

For example, should a fn like .sum_all, that just has a loop, that does not use `[]`, produce a warning/notice, if it is tagged with `[direct_array_access]`?

I can see value in both.

**Comment by penguindark** [comment]
On this topic I've not a strong opinion.
A warning may be helpful, but IMHO it can work even without warning.
